include/core/checkpoint_control.h

```diff
@@ -0,0 +1,91 @@
+// SPDX-FileCopyrightText: Copyright (c) 2020-2024 The Lethe Authors
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR LGPL-2.1-or-later
+
+#ifndef lethe_checkpoint_control_h
+#define lethe_checkpoint_control_h
+
+#include <boost/archive/text_iarchive.hpp>
+#include <boost/archive/text_oarchive.hpp>
+
+class CheckpointControl
+{
+public:
+  /**
+   * @brief The CheckpointControl is responsible for keeping track of the
+   * next checkpoint id to use.
+   *
+   * @param checkpoint_freq Checkpoint frequency.
+   * @param restart_filename Filename used for the prefix during checkpointing.
+   */
+  CheckpointControl(const unsigned int checkpoint_freq,
+                    const std::string &restart_filename)
+    : max_checkpoint_id(2)
+    , next_checkpoint_id(max_checkpoint_id)
+    , checkpointing_frequency(checkpoint_freq)
+    , filename(restart_filename)
+  {}
+
+  /**
+   * @brief Return true if a checkpoint should be made at the current time step.
+   * @param time_step_number Current time step.
+   */
+  bool
+  is_checkpoint_time_step(const unsigned int time_step_number)
+  {
+    return (time_step_number % checkpointing_frequency) == 0;
+  }
+
+  /**
+   * @brief Calculate the next checkpoint id.
+   */
+  void
+  increment_checkpoint_id()
+  {
+    next_checkpoint_id = (next_checkpoint_id + 1) % max_checkpoint_id;
+  }
```

Comment:
Where is this function called? It does not seem to be called from the checkpoint controller. Don,t forget that in the end we can make sure that the checkpoint controller is the only one that increments the checkpoint id.

Reply:
It's called in dem.cc when you enter the of statement